src/pudl/transform/ferc1.py

```diff
@@ -6745,31 +6752,100 @@ def make_calculation_dimensions_explicit(
-                    "xbrl_factoid": "xbrl_factoid_calc",
-                    "table_name": "table_name_calc",
-                }
-            )
-        )
+        observed_dim = table_dimensions_ferc1[
+            on_cols + [dim_col]
+        ].drop_duplicates()  # bc there are dupes after we removed the other dim cols
         null_dim_mask = calc_comps_w_dims[dim_col].isnull()
         null_dim = calc_comps_w_dims[null_dim_mask].drop(columns=[dim_col])
         calc_comps_w_implied_dims = pd.merge(
             null_dim,
             observed_dim,
-            on=["table_name_calc", "xbrl_factoid_calc"],
+            on=on_cols,
             how="left",
         )
         calc_comps_w_explicit_dims = calc_comps_w_dims[~null_dim_mask]
         calc_comps_w_dims = pd.concat(
             [calc_comps_w_implied_dims, calc_comps_w_explicit_dims]
         )
     return calc_comps_w_dims
+
+
+def add_parent_dimensions(calc_comps, dimensions):
+    """Add in the dimensions of the parents in the calculation components.
+
+    Together :meth:`process_xbrl_metadata_calculations` and
+    :func:`make_calculation_dimensions_explicit` generates calculation component table
+    with pk columns of: ``table_name_parent``, ``xbrl_factoid_parent``, ``table_name``,
+    ``xbrl_factoid``. The table also contains dimension columns (ex: ``utility_type``,
+    ``plant_function``, etc.) which indicate the dimensions of the calculation component
+    factoid - not the parent factoid. This function attempts to add in additional
+    ``dimension`` columns for the parent-side of the calculation.
+
+    First, we treat the totals in these dimension columns by assuming that any "total"
+    we observe (i.e. ``utility_type=="total"``) can be summed up  of all non-total
+    values within that same dimension (i.e.
+    ``utility_type.isin(["electric", "gas", "other"])``). We implement this by creating
+    new calculation component records which have a parent of dimension "total" and
+    calculation components of all observed non-total dimension values. This is done by
+    a broadcast merge of the non-total dimension calculation component values onto
+    parent total records. The result of this is the creation of new records which
+    communicate this total -> non-total calculation that was not previously encapsulated
+    within the calculation components.
+
+    Then, we add parent-level dimension values for all of the original calculation
+    component records in ``calc_comps``. We assume that every parent factoid should have
+    the same dimension values as its calculation component dimension values.
+
+    Args:
+        calc_comps: a table of calculation component records which have had some manual
+            calculation fixes applied.
+        dimensions: list of dimension columns to check.
+    """
+    calc_comp_idx = [
+        "table_name_parent",
+        "xbrl_factoid_parent",
+        "table_name",
+        "xbrl_factoid",
+    ]
```

Comment:
Do the PKs necessarily not include the dimension columns? E.g. couldn't there be cases where we define different calculations depending `utility_type`? This almost happens where we have one calculation with lots of details from the `plant_in_service_ferc1` table, and a different simpler calculation for the gas and other utility types. In that case I think it happens to be the case that the calculation components are being reported in different tables, but does that necessarily have to be true?

Or is it the case that at this point in the processing of the calculation components table we're still adding values to the dimension columns, and at the end of this step they're going to be part of the PKs in a way that they weren't before? That seems true with respect to the addition of the `total` calculations, but is it true more generally?